### PR TITLE
FIX: extends typing-extensions to be installed with python 3.8+ #21566

### LINKED ISSUE #21566: typing_extensions package isn't installed with apache-airflow-providers-amazon causing an issue for SqlToS3Operator
### Apache Airflow Provider(s)

amazon

### Versions of Apache Airflow Providers

3.0.0rc2

### Apache Airflow version

2.2.3 (latest released)

### Python version

Python 3.9.7 (default, Oct 12 2021, 02:43:43)

### Operating System

Debian GNU/Linux 11 (bullseye)

### Deployment

Docker-Compose

### Deployment details

_No response_

### What happened

I was working on adding this operator to a DAG and it failed to import due to a lack of a required file

### What you expected to happen

_No response_

### How to reproduce

Add 

```

from airflow.providers.amazon.aws.transfers.sql_to_s3 import SqlToS3Operator
```

to a dag

### Anything else

This can be resolved by adding `typing-extensions==4.1.1` to `requirements.txt` when building the project (locally)

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by frankcash**
~~Uses built in util to handle import of literal until python 3.8 is phased out.~~

Changes `setup.cfg` to install `typing-extensions` will all versions of python, not just `<3.8`

closes: #21566 

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/providers/amazon/aws/sensors/sqs.py [hunk 1]
```diff
@@ -21,11 +21,11 @@
 from typing import TYPE_CHECKING, Any, Optional, Sequence
 
 from jsonpath_ng import parse
-from typing_extensions import Literal
 
 from airflow.exceptions import AirflowException
 from airflow.providers.amazon.aws.hooks.sqs import SqsHook
 from airflow.sensors.base import BaseSensorOperator
+from airflow.typing_compat import Literal
```
Comment (ashb) [inline airflow/providers/amazon/aws/sensors/sqs.py]: We'll need to check airflow versions here.

What is the minumum version that the provider now needs, and when did we release a version of Airflow core that contained the "Literal" compat.

Comment (ashb) [reply airflow/providers/amazon/aws/sensors/sqs.py]: https://github.com/apache/airflow/pull/16582 - in 2.1.1 is when this was added.

Comment (frankcash) [reply airflow/providers/amazon/aws/sensors/sqs.py]: Minimum version appears to be 2.1.0  for the provider https://github.com/apache/airflow/blob/main/airflow/providers/amazon/provider.yaml#L40

**Comment by uranusjr** [comment]
Why not just add typing_extension as a dependency?

File: airflow/providers/amazon/aws/sensors/sqs.py (hunk 1 above)
Comment (potiuk) [reply airflow/providers/amazon/aws/sensors/sqs.py]: Then we can't use it, I am afraid.

Comment (potiuk) [reply airflow/providers/amazon/aws/sensors/sqs.py]: The eror is clear about it https://github.com/apache/airflow/runs/5191554253?check_suite_focus=true

Comment (frankcash) [reply airflow/providers/amazon/aws/sensors/sqs.py]: Sounds good, thanks for the feedback. Is there a preferred way I should go about solving this @potiuk ?

Comment (frankcash) [reply airflow/providers/amazon/aws/sensors/sqs.py]: As @uranusjr pointed out we could change the line in setup.cfg that doesn't install the lib if on python 3.8+ 

I included that as a potential fix in the issue https://github.com/apache/airflow/issues/21566#issuecomment-1039537580

Comment (potiuk) [reply airflow/providers/amazon/aws/sensors/sqs.py]: I am for this solution as well:

```
typing-extensions>=3.7.4;python_version<"3.8" 
```

File: setup.cfg [hunk 2]
```diff
@@ -137,7 +137,7 @@ install_requires =
     tabulate>=0.7.5, <0.9
     tenacity>=6.2.0
     termcolor>=1.1.0
-    typing-extensions>=3.7.4;python_version<"3.8"
```
Comment (potiuk) [inline setup.cfg:166]: I think we should add a comment here that we can remove this dependency when one of two things happen:

1) We drop Pyhon 3.7
2) We add min airflow version for all providers to be 2.2 - in which case we can replace it with compat.

Comment (potiuk) [reply setup.cfg:166]: BTW. The 2. might happen in a few months, so it's good to note it to remember and remove it eventually.

Comment (frankcash) [reply setup.cfg:166]: Addressed in latest commit

**Review by uranusjr — APPROVED** [review]

**Comment by frankcash** [comment]
Looks like the Github Actions are "failed" because one was cancelled? `Error: The operation was canceled.`

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Closed/reopened to check if this was a flaky fail.

**Comment by frankcash** [comment]
> It could be that the typing extension actually causes some too long running constraint generation.

looks like it timed out again

**Comment by frankcash** [comment]
@potiuk  / @uranusjr  I've re-run this a few times but it continues to time out at the Screen Shot 2022-02-16 at 9 32 39 AM step

I am confused on what I need to change / do to get this to resolve

**Comment by potiuk** [comment]
> I am confused on what I need to change / do to get this to resolve -- would someone point me in the right direction?

Yeah. I was afraid it is not "accident". I will have to take a look at that more closely, but I think the problem is that it creates unsolvable conflict with one of the relased providers (we prepare few variants of the constraints and apparently this one gets triggered. 

What happens under the hood `pip` gets into backtracking and tries to find a solution to impossible conflict resolution and it takes awfully lot of time to resolve it. I am afraid this one will need to wait until we take a bit closer look what's going on in detail. I should find time for it this week.

**Comment by frankcash** [comment]
Sounds good, thank you so much for your time

**Comment by potiuk** [comment]
> Sounds good, thank you so much for your time

I think it should work fine now. We released new providers, and increased timeout for constraint generation, so it should work as both potential reasons for the hanging should have been addressed. Rebased and let's see if it succeeds.

**Review by potiuk — APPROVED** [review]

**Comment by frankcash** [comment]
Looks like the provider tests failed with the rebase `Tests / MSSQL2019-latest, Py3.7: Always API Core Other CLI Providers WWW Integration (pull_request) Failing after 19m — MSSQL2019-latest, Py3.7: Always API Core Other CLI Providers WWW Integration`

**Comment by frankcash** [comment]
Looks like the failing tests will be fixed with https://github.com/apache/airflow/pull/21733

**Comment by frankcash** [comment]
@potiuk / @uranusjr lgtm if one of ya'll want to merge it

**Comment by frankcash** [comment]
Keeps timing out at `Build Images / Build CI image 3.7 (pull_request_target) In progress — This check has started...` 🙃

**Comment by frankcash** [comment]
it looks like a lot of other builds are failing at the build images / build ci image {python_version} step..

**Comment by potiuk** [comment]
> it looks like a lot of other builds are failing at the build images / build ci image {python_version} step..

Not all - only those that change "setup.py". This is pip getting into backtracking loop.

**Comment by potiuk** [comment]
@frankcash ->  I would really love if you could upvote and comment here: https://github.com/pypa/pip/pull/10258#issuecomment-1025125753. 

The problem is that we have no way currently to determine what is the root cause of the problem. The problem is that `pip` is not able to solve dependencies in a reasonable time and runs backtracking for hours - but it does not really give us the possibilty to get the root cause message that could allow us to determine how we could help `pip`. 

I proposed to add optional timeout on `pip install` execution producing a "conflict" message if the timeout is reached. That could  give us at least a chance to understand what is the problem (someone released a new dependency that makes `pip` fail installing airflow from the scratch and we need to help `pip` to determine the right set of dependencies because it is unable to figure it out on its own).

**Comment by frankcash** [comment]
Thanks for linking that Jarek, I will provide a comment on it.

**Comment by potiuk** [comment]
@frankcash I think I managed to workaround the pip resolver issue with https://github.com/apache/airflow/pull/21824 - please rebase to latest main.

**Comment by frankcash** [comment]
🤞

**Comment by frankcash** [comment]
@potiuk it worked yay

**Pull request merged**